File: node/10_search_query_advanced_ops.js
const fs = require("fs")
const { MilvusClient, DataType, sleep } = require("@zilliz/milvus2-sdk-node")

const address = "YOUR_CLUSTER_ENDPOINT"
const token = "YOUR_CLUSTER_TOKEN"
const collectionName = "medium_articles_2020"
const data_file = `./medium_articles_2020_dpr.json`

async function main() {
    // 1. Connect to the cluster
    const client = new MilvusClient({address, token})
    
    // 2. Define fields
    fields = [
        {
            name: "id",
            data_type: DataType.Int64,
            is_primary_key: true,
            auto_id: true
        },
        {
            name: "title",
            data_type: DataType.VarChar,
            max_length: 512
        },
        {
            name: "title_vector",
            data_type: DataType.FloatVector,
            dim: 768
        },
        {
            name: "article_meta",
            // This field is a JSON field.
            data_type: DataType.JSON
        }
    ]
    
    // 3. Create collection
    res = await client.createCollection({
        collection_name: collectionName,
        fields: fields
    })

    console.log(res)

    // Output
    // 
    // { error_code: 'Success', reason: '', code: 0 }
    // 


    // 4. Create index
    res = await client.createIndex({
        collection_name: collectionName,
        field_name: "title_vector",
        index_type: "IVF_FLAT",
        metric_type: "L2",
        params: {
            nlist: 16384
        }
    })

    console.log(res)

    // Output
    // 
    // { error_code: 'Success', reason: '', code: 0 }
    // 


    res = await client.loadCollection({
        collection_name: collectionName
    })

    console.log(res)

    // Output
    // 
    // { error_code: 'Success', reason: '', code: 0 }
    // 
    

    // 5. Insert vectors
    
    const data = JSON.parse(fs.readFileSync(data_file))

    // read rows
    const rows = data["rows"]
    rows.forEach(row => {
        // add a new field in the row object
        row.article_meta = {}
        // move the following fields into the new field
        row.article_meta.link = row.link
        delete row.link
        row.article_meta.reading_time = row.reading_time
        delete row.reading_time
        row.article_meta.publication = row.publication
        delete row.publication
        row.article_meta.claps = row.claps
        delete row.claps
        row.article_meta.responses = row.responses
        delete row.responses
        row.article_meta.tags_1 = Array.from({length: 40}, () => Math.floor(Math.random() * 40));   
        row.article_meta.tags_2 = Array.from({length: 10}, () => Array.from({length: 4}, () => Math.floor(Math.random() * 40)))      
    });

    console.log(Object.keys(rows[0]))

    // Output
    // 
    // [ 'id', 'title', 'title_vector', 'article_meta' ]
    // 


    //insert vectors
    res = await client.insert({
        collection_name: collectionName,
        data: rows
    })

    console.log(res)

    // Output
    // 
    // {
    //   succ_index: [
    //      0,  1,  2,  3,  4,  5,  6,  7,  8,  9, 10, 11,
    //     12, 13, 14, 15, 16, 17, 18, 19, 20, 21, 22, 23,
    //     24, 25, 26, 27, 28, 29, 30, 31, 32, 33, 34, 35,
    //     36, 37, 38, 39, 40, 41, 42, 43, 44, 45, 46, 47,
    //     48, 49, 50, 51, 52, 53, 54, 55, 56, 57, 58, 59,
    //     60, 61, 62, 63, 64, 65, 66, 67, 68, 69, 70, 71,
    //     72, 73, 74, 75, 76, 77, 78, 79, 80, 81, 82, 83,
    //     84, 85, 86, 87, 88, 89, 90, 91, 92, 93, 94, 95,
    //     96, 97, 98, 99,
    //     ... 5879 more items
    //   ],
    //   err_index: [],
    //   status: { error_code: 'Success', reason: '', code: 0 },
    //   IDs: { int_id: { data: [Array] }, id_field: 'int_id' },
    //   acknowledged: false,
    //   insert_cnt: '5979',
    //   delete_cnt: '0',
    //   upsert_cnt: '0',
    //   timestamp: '445318026051190787'
    // }
    // 


    await sleep(5000)

    // 6. Count entities

    counts = await client.query({
        collection_name: collectionName,
        filter: "",
        output_fields: ["count(*)"]
    })

    console.log(counts)

    // Output
    // 
    // {
    //   status: { error_code: 'Success', reason: '', code: 0 },
    //   data: [ { 'count(*)': '5979' } ]
    // }
    // 

    // 7. Count entities with condition

    // matches all articles with tags_1 having the member 16
    const expr_1 = 'JSON_CONTAINS(article_meta["tags_1"], 16)'

    // matches all articles with tags_2 having the member [5, 3, 39, 8]
    const expr_2 = 'JSON_CONTAINS(article_meta["tags_2"], [5, 3, 39, 8])'

    // matches all articles with tags_1 having a member from [5, 3, 39, 8]
    const expr_3 = 'JSON_CONTAINS_ANY(article_meta["tags_1"], [5, 3, 39, 8])'

    // matches all articles with tags_1 having all members from [2, 4, 6]
    const expr_4 = 'JSON_CONTAINS_ALL(article_meta["tags_1"], [2, 4, 6])'

    counts = await client.query({
        collection_name: collectionName,
        filter: expr_1,
        output_fields: ["count(*)"]
    })

    console.log(counts)

    // Output
    // 
    // {
    //   status: { error_code: 'Success', reason: '', code: 0 },
    //   data: [ { 'count(*)': '3825' } ]
    // }
    // 

    // 8. Check if a specific element exists in JSON field

    // Search

    res = await client.search({
        collection_name: collectionName,
        vector: rows[0].title_vector,
        limit: 5,
        filter: expr_3,
        output_fields: ["title", "article_meta"]
    })

    console.log(res)

    // Output
    // 
    // {
    //   status: { error_code: 'Success', reason: '', code: 0 },
    //   results: [
    //     {
    //       score: 0,
    //       id: '0',
    //       title: 'The Reported Mortality Rate of Coronavirus Is Not Important',
    //       article_meta: [Object]
    //     },
    //     {
    //       score: 0.29999834299087524,
    //       id: '3177',
    //       title: 'Following the Spread of Coronavirus',
    //       article_meta: [Object]
    //     },
    //     {
    //       score: 0.36103832721710205,
    //       id: '5607',
    //       title: 'The Hidden Side Effect of the Coronavirus',
    //       article_meta: [Object]
    //     },
    //     {
    //       score: 0.37674015760421753,
    //       id: '5641',
    //       title: 'Why The Coronavirus Mortality Rate is Misleading',
    //       article_meta: [Object]
    //     },
    //     {
    //       score: 0.416297972202301,
    //       id: '3441',
    //       title: 'Coronavirus shows what ethical Amazon could look like',
    //       article_meta: [Object]
    //     }
    //   ]
    // }
    // 

    tags_1 = res.results.map((item) => item.article_meta["tags_1"])

    console.log(tags_1)

    // Output
    // 
    // [
    //   [
    //     24, 20, 37, 11,  3, 16, 26, 39, 18, 34,
    //     13, 25,  9, 37, 29, 30, 18, 34, 27, 22,
    //      6, 23,  3,  5, 14,  0, 23,  2, 36, 37,
    //     31, 28,  7, 18, 16,  1,  9, 12, 17,  3
    //   ],
    //   [
    //      9,  6, 34, 16,  2,  5, 15, 36,  6, 37,
    //     17,  7,  7, 13, 21,  6, 33, 37, 16, 24,
    //     35, 32, 17, 11, 33, 16, 22, 22, 39, 36,
    //      7, 37, 15, 21, 31, 16,  8, 37, 18, 17
    //   ],
    //   [
    //     23, 10, 35, 32, 23,  1,  1, 31, 10, 20,
    //     20,  2, 15,  3,  3, 22,  8, 27, 24, 18,
    //     36, 24, 22,  1, 24, 31,  9, 20, 25, 22,
    //     10, 36, 29, 25,  3, 17, 37,  2,  6,  0
    //   ],
    //   [
    //     37, 22, 38, 32, 20, 27,  3, 28, 23, 26,
    //     23, 22, 24,  7, 38, 23,  2,  5, 19,  6,
    //     14, 38, 24, 27, 11,  9, 30, 18, 38,  2,
    //     19, 16,  9,  1, 12, 16,  4,  0,  5,  2
    //   ],
    //   [
    //     18, 38, 28, 36,  0, 31, 31, 39,  1, 21,
    //     18,  0, 18, 38, 20, 31, 17,  1, 37,  7,
    //     29, 33, 13, 32, 29, 27, 27, 39, 19, 36,
    //     30,  4, 21,  6, 31, 26, 35,  7, 11, 24
    //   ]
    // ]
    // 

    // Query

    res = await client.query({
        collection_name: collectionName,
        filter: expr_4,
        limit: 5,
        output_fields: ["title", "article_meta"]
    })

    console.log(res)

    // Output
    // 
    // {
    //   status: { error_code: 'Success', reason: '', code: 0 },
    //   data: [
    //     {
    //       title: 'Maternity leave shouldn’t set women back',
    //       article_meta: [Object],
    //       id: '3'
    //     },
    //     {
    //       title: 'Python NLP Tutorial: Information Extraction and Knowledge Graphs',
    //       article_meta: [Object],
    //       id: '4'
    //     },
    //     {
    //       title: 'Building Comprehensible Customer Churn Prediction Models',
    //       article_meta: [Object],
    //       id: '7'
    //     },
    //     {
    //       title: '5 Underrated Traits That Separate Exceptional Teams From the Rest',
    //       article_meta: [Object],
    //       id: '8'
    //     },
    //     {
    //       title: 'Building high performance startup teams',
    //       article_meta: [Object],
    //       id: '10'
    //     }
    //   ]
    // }
    // 

    tags_1 = res.data.map((item) => item.article_meta["tags_1"])

    console.log(tags_1)

    // Output
    // 
    // [
    //   [
    //      2, 34, 25, 19, 19,  6,  3, 37, 16,  1,
    //     27,  9,  9, 26, 27, 34, 30, 12, 11, 36,
    //     30,  3,  1,  2, 24,  4, 14, 33, 32, 14,
    //     26, 12, 22, 28, 10,  6,  2, 39, 10, 34
    //   ],
    //   [
    //     37,  3, 11, 27, 20,  6,  4, 33,  7,  4,
    //     37, 39, 29, 19, 35, 24, 29, 25, 38, 25,
    //      7, 27,  8, 19, 27,  7, 15, 36,  8, 28,
    //     29, 13, 13, 30, 29, 30, 22,  2, 35, 30
    //   ],
    //   [
    //      4, 26,  1, 31, 10,  5, 33, 8,  0,  2,
    //     39, 37, 38, 25, 12, 28, 19, 9, 13, 29,
    //     22, 19, 24, 25,  3, 34, 22, 3,  9,  3,
    //      8, 27, 36, 26,  5, 20, 18, 6, 26, 30
    //   ],
    //   [
    //     27,  1,  5, 27, 14,  4,  2, 35,  6, 35,
    //     34,  0, 22, 37, 33,  4,  0, 26,  7,  1,
    //     22, 21,  8, 28, 31, 24, 33, 22, 27, 22,
    //     19, 20,  3, 34,  8, 13, 24, 28, 25, 34
    //   ],
    //   [
    //     38, 12,  4, 18, 17,  2, 25, 31,  7, 23,
    //      0, 10, 23, 35, 37, 32, 26,  9, 32, 27,
    //      8, 38,  5, 30, 10, 10,  7, 25, 29, 10,
    //     27, 23, 34, 27, 38,  0,  6, 38, 34, 23
    //   ]
    // ]
    // 

    // 7. Get collection info

    res = await client.describeCollection({
        collection_name: collectionName
    })

    console.log(res);

    // Output
    // 
    // {
    //   virtual_channel_names: [ 'by-dev-rootcoord-dml_11_445311585783000491v0' ],
    //   physical_channel_names: [ 'by-dev-rootcoord-dml_11' ],
    //   aliases: [],
    //   start_positions: [],
    //   properties: [],
    //   status: { error_code: 'Success', reason: '', code: 0 },
    //   schema: {
    //     fields: [ [Object], [Object], [Object], [Object] ],
    //     name: 'medium_articles_2020',
    //     description: '',
    //     autoID: false,
    //     enable_dynamic_field: false
    //   },
    //   collectionID: '445311585783000491',
    //   created_timestamp: '445318018907504647',
    //   created_utc_timestamp: '1698753429060',
    //   shards_num: 1,
    //   consistency_level: 'Bounded',
    //   collection_name: 'medium_articles_2020',
    //   db_name: 'default',
    //   num_partitions: '1'
    // }
    // 


    // 8. Drop collection

    res = await client.dropCollection({
        collection_name: collectionName
    })     
    
    console.log(res);

    // Output
    // 
    // { error_code: 'Success', reason: '', code: 0 }
    // 

}

main()
    

File: python/10_search_query_advanced_ops.py
import json, os, time, random
from contextlib import suppress
from pymilvus import connections, FieldSchema, CollectionSchema, DataType, Collection, utility

CLUSTER_ENDPOINT="YOUR_CLUSTER_ENDPOINT" # Set your cluster endpoint
TOKEN="YOUR_CLUSTER_TOKEN" # Set your token
COLLECTION_NAME="medium_articles_2020" # Set your collection name
DATASET_PATH="{}/../medium_articles_2020_dpr.json".format(os.path.dirname(__file__)) # Set your dataset path

# 0. Connect to cluster
connections.connect(
    uri=CLUSTER_ENDPOINT, # Public endpoint obtained from Zilliz Cloud
    token=TOKEN, # API key or a colon-separated cluster username and password
)

# 1. Define fields
fields = [
    FieldSchema(name="id", dtype=DataType.INT64, is_primary=True, auto_id=True, max_length=100),
    FieldSchema(name="title", dtype=DataType.VARCHAR, max_length=512),
    FieldSchema(name="title_vector", dtype=DataType.FLOAT_VECTOR, dim=768),
    # The following field is a JSON field
    FieldSchema(name="article_meta", dtype=DataType.JSON)
]

# 2. Build the schema
schema = CollectionSchema(
    fields,
    description="Schema of Medium articles",
    enable_dynamic_field=False
)

# 3. Create collection
collection = Collection(
    name=COLLECTION_NAME, 
    description="Medium articles published between Jan and August in 2020 in prominent publications",
    schema=schema
)

# 4. Index collection
# 'index_type' defines the index algorithm to be used.
#    AUTOINDEX is the only option.
#
# 'metric_type' defines the way to measure the distance 
#    between vectors. Possible values are L2, IP, and Cosine,
#    and defaults to Cosine.
index_params = {
    "index_type": "AUTOINDEX",
    "metric_type": "L2",
    "params": {}
}

# To name the index, do as follows:
collection.create_index(
    field_name="title_vector", 
    index_params=index_params,
)

# 5. Load collection
collection.load()

# Get loading progress
progress = utility.loading_progress(COLLECTION_NAME)

print(progress)

# Output
#
# {
#     "loading_progress": "100%"
# }



# 6. Prepare data

# Prepare a list of rows
with open(DATASET_PATH) as f:
    data = json.load(f)
    list_of_rows = data['rows']

    rows = []
    for row in list_of_rows:
        # Remove the id field because auto-id is enabled for the primary key
        del row['id']
        # Create the article_meta field and 
        row['article_meta'] = {}
        # Move the following keys into the article_meta field
        row['article_meta']['link'] = row.pop('link')
        row['article_meta']['reading_time'] = row.pop('reading_time')
        row['article_meta']['publication'] = row.pop('publication')
        row['article_meta']['claps'] = row.pop('claps')
        row['article_meta']['responses'] = row.pop('responses')
        row['article_meta']['tags_1'] = [ random.randint(0, 40) for x in range(40)]
        row['article_meta']['tags_2'] = [ [ random.randint(0, 40) for y in range(4) ] for x in range(10) ]
        # Append this row to the data_rows list
        rows.append(row)

print(rows[:1])

# Output
#
# [
#     {
#         "title": "The Reported Mortality Rate of Coronavirus Is Not Important",
#         "title_vector": [
#             0.041732933,
#             0.013779674,
#             -0.027564144,
#             -0.013061441,
#             0.009748648,
#             0.00082446384,
#             -0.00071647146,
#             0.048612226,
#             -0.04836573,
#             -0.04567751,
#             "(758 more items hidden)"
#         ],
#         "article_meta": {
#             "link": "https://medium.com/swlh/the-reported-mortality-rate-of-coronavirus-is-not-important-369989c8d912",
#             "reading_time": 13,
#             "publication": "The Startup",
#             "claps": 1100,
#             "responses": 18,
#             "tags_1": [
#                 20,
#                 39,
#                 4,
#                 4,
#                 11,
#                 30,
#                 6,
#                 23,
#                 29,
#                 1,
#                 "(30 more items hidden)"
#             ],
#             "tags_2": [
#                 [
#                     37,
#                     29,
#                     23,
#                     27
#                 ],
#                 [
#                     6,
#                     34,
#                     17,
#                     33
#                 ],
#                 [
#                     39,
#                     9,
#                     15,
#                     37
#                 ],
#                 [
#                     3,
#                     28,
#                     21,
#                     10
#                 ],
#                 [
#                     40,
#                     15,
#                     36,
#                     20
#                 ],
#                 [
#                     15,
#                     24,
#                     40,
#                     26
#                 ],
#                 [
#                     8,
#                     36,
#                     18,
#                     19
#                 ],
#                 [
#                     36,
#                     21,
#                     16,
#                     28
#                 ],
#                 [
#                     14,
#                     6,
#                     32,
#                     18
#                 ],
#                 [
#                     23,
#                     31,
#                     36,
#                     39
#                 ]
#             ]
#         }
#     }
# ]



# 7. Insert data
results = collection.insert(rows)

print(f"Data inserted successfully! inserted rows: {results.insert_count}")

# Output
#
# Data inserted successfully! inserted rows: 5979



time.sleep(10)

# 8. Count entities

counts = collection.query(expr="", output_fields=["count(*)"])

print(counts)

# Output
#
# [
#     {
#         "count(*)": 5979
#     }
# ]



# 9. Count entities with condition

expr = 'article_meta["claps"] > 30 and article_meta["reading_time"] < 10'

counts = collection.query(expr=expr, output_fields=["count(*)"])

print(counts)

# Output
#
# [
#     {
#         "count(*)": 4304
#     }
# ]



# 10. Check if a specific element exists in a JSON field

# Search

# matches all articles with tags_1 having the member 16
expr_1 = 'JSON_CONTAINS(article_meta["tags_1"], 16)'

# matches all articles with tags_2 having the member [5, 3, 39, 8]
expr_2 = 'JSON_CONTAINS(article_meta["tags_2"], [5, 3, 39, 8])'

# matches all articles with tags_1 having a member from [5, 3, 39, 8]
expr_3 = 'JSON_CONTAINS_ANY(article_meta["tags_1"], [5, 3, 39, 8])'

# matches all articles with tags_1 having all members from [2, 4, 6]
expr_4 = 'JSON_CONTAINS_ALL(article_meta["tags_1"], [2, 4, 6])'

query_vector = rows[0]['title_vector']

# Define search parameters
search_params = {
    "metric_type": "L2",
    "params": {"nprobe": 10}
}

res = collection.search(
    data=[query_vector],
    anns_field="title_vector",
    param={"metric_type": "L2", "params": {"nprobe": 10}},
    limit=5,
    expr=expr_1,
    output_fields=["title", "article_meta"]
)

ids = [ hits.ids for hits in res ]

print(ids)

# Output
#
# [
#     [
#         445311585782930475,
#         445311585782930509,
#         445311585782928309,
#         445311585782929013,
#         445311585782925710
#     ]
# ]



distances = [ hits.distances for hits in res ]

print(distances)

# Output
#
# [
#     [
#         0.36103832721710205,
#         0.37674015760421753,
#         0.416297972202301,
#         0.4928317368030548,
#         0.49443864822387695
#     ]
# ]



def get_tags_1(value, target):
    try:
        return value.index(target) >= 0
    except (ValueError):
        return False

results = [ {
    "id": hit.id,
    "distance": hit.distance,
    "entity": {
        "title": hit.entity.get("title"),
        "link": hit.entity.get("article_meta")['link'],
        "tags_1": get_tags_1(hit.entity.get("article_meta")["tags_1"], 16),
    }
} for hits in res for hit in hits ]

print(results)

# Output
#
# [
#     {
#         "id": 445311585782930475,
#         "distance": 0.36103832721710205,
#         "entity": {
#             "title": "The Hidden Side Effect of the Coronavirus",
#             "link": "https://medium.com/swlh/the-hidden-side-effect-of-the-coronavirus-b6a7a5ee9586",
#             "tags_1": true
#         }
#     },
#     {
#         "id": 445311585782930509,
#         "distance": 0.37674015760421753,
#         "entity": {
#             "title": "Why The Coronavirus Mortality Rate is Misleading",
#             "link": "https://towardsdatascience.com/why-the-coronavirus-mortality-rate-is-misleading-cc63f571b6a6",
#             "tags_1": true
#         }
#     },
#     {
#         "id": 445311585782928309,
#         "distance": 0.416297972202301,
#         "entity": {
#             "title": "Coronavirus shows what ethical Amazon could look like",
#             "link": "https://medium.com/swlh/coronavirus-shows-what-ethical-amazon-could-look-like-7c80baf2c663",
#             "tags_1": true
#         }
#     },
#     {
#         "id": 445311585782929013,
#         "distance": 0.4928317368030548,
#         "entity": {
#             "title": "Will Coronavirus Impact Freelancers\u2019 Ability to Rent?",
#             "link": "https://medium.com/swlh/will-coronavirus-impact-freelancers-ability-to-rent-ae11f2847bab",
#             "tags_1": true
#         }
#     },
#     {
#         "id": 445311585782925710,
#         "distance": 0.49443864822387695,
#         "entity": {
#             "title": "Choosing the right performance metrics can save lives against Coronavirus",
#             "link": "https://towardsdatascience.com/choosing-the-right-performance-metrics-can-save-lives-against-coronavirus-2f27492f6638",
#             "tags_1": true
#         }
#     }
# ]



# query

res = collection.query(
    limit=5,
    expr=expr_4,
    output_fields=["title", "article_meta"]
)

res = [ {
    "title": x.get("title"),
    "article_meta": {
        "link": x.get("article_meta")['link'],
        "reading_time": x.get("article_meta")['reading_time'],
        "publication": x.get("article_meta")['publication'],
        "claps": x.get("article_meta")['claps'],
        "responses": x.get("article_meta")['responses'],
        "tags_1": get_tags_1(x.get("article_meta")['tags_1'], 2),
    },
} for x in res ]

print(res)

# Output
#
# [
#     {
#         "title": "The Reported Mortality Rate of Coronavirus Is Not Important",
#         "article_meta": {
#             "link": "https://medium.com/swlh/the-reported-mortality-rate-of-coronavirus-is-not-important-369989c8d912",
#             "reading_time": 13,
#             "publication": "The Startup",
#             "claps": 1100,
#             "responses": 18,
#             "tags_1": true
#         }
#     },
#     {
#         "title": "Dashboards in Python: 3 Advanced Examples for Dash Beginners and Everyone Else",
#         "article_meta": {
#             "link": "https://medium.com/swlh/dashboards-in-python-3-advanced-examples-for-dash-beginners-and-everyone-else-b1daf4e2ec0a",
#             "reading_time": 14,
#             "publication": "The Startup",
#             "claps": 726,
#             "responses": 3,
#             "tags_1": true
#         }
#     },
#     {
#         "title": "Would you rather have 8% of $25 or 25% of $8?",
#         "article_meta": {
#             "link": "https://medium.com/swlh/would-you-rather-have-8-of-25-or-25-of-8-486b3bc48f28",
#             "reading_time": 3,
#             "publication": "The Startup",
#             "claps": 208,
#             "responses": 5,
#             "tags_1": true
#         }
#     },
#     {
#         "title": "Big Reasons for Expanding Your Vocabulary, and How to Do It.",
#         "article_meta": {
#             "link": "https://medium.com/swlh/big-reasons-for-expanding-your-vocabulary-and-how-to-do-it-c78766e4df2",
#             "reading_time": 8,
#             "publication": "The Startup",
#             "claps": 417,
#             "responses": 3,
#             "tags_1": true
#         }
#     },
#     {
#         "title": "How to Find Things and Do Research in a Discovery Phase",
#         "article_meta": {
#             "link": "https://medium.com/swlh/how-to-find-things-and-do-research-in-a-discovery-phase-6c3e4f50b208",
#             "reading_time": 5,
#             "publication": "The Startup",
#             "claps": 330,
#             "responses": 0,
#             "tags_1": true
#         }
#     }
# ]



# 11. Drop collection

utility.drop_collection(COLLECTION_NAME)